Foundation/URLCache.swift

```diff
@@ -193,7 +326,123 @@ open class URLCache : NSObject {
+                
+                cacheDirectory = url
+            } catch {
+                cacheDirectory = nil
+            }
+        }
+    }
+    
+    private func identifier(for request: URLRequest) -> String? {
+        guard let url = request.url else { return nil }
+        
+        if let host = url.host {
+            var data = Data()
+            data.append(Data(host.lowercased(with: NSLocale.system).utf8))
+            data.append(0)
+            let port = url.port ?? -1
+            data.append(Data("\(port)".utf8))
+            data.append(0)
+            data.append(Data(url.path.utf8))
+            data.append(0)
+        
+            return data.base64EncodedString()
+        } else {
+            return nil
+        }
+    }
+    
+    private struct DiskEntry {
+        static let pathExtension = "storedcachedurlresponse"
+        
+        var url: URL
+        var date: Date
+        var identifier: String
+        
+        init?(_ url: URL) {
+            if url.pathExtension.localizedCompare(DiskEntry.pathExtension) != .orderedSame {
+                return nil
+            }
+            
+            let parts = url.deletingPathExtension().lastPathComponent.components(separatedBy: ".")
+            guard parts.count == 2 else { return nil }
+            let (timeString, identifier) = (parts[0], parts[1])
+            
+            guard let time = Int64(timeString) else { return nil }
+            
+            self.date = Date(timeIntervalSinceReferenceDate: TimeInterval(time))
+            self.identifier = identifier
+            self.url = url
+        }
+    }
+    
+    private func enumerateDiskEntries(includingPropertiesForKeys keys: [URLResourceKey] = [], using block: (DiskEntry, inout Bool) -> Void) {
+        guard let directory = cacheDirectory else { return }
+        for url in (try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: keys)) ?? [] {
+            if let entry = DiskEntry(url) {
+                var stop = false
+                block(entry, &stop)
+                if stop {
+                    return
+                }
+            }
+        }
+    }
+    
+    private func diskContentsURL(for request: URLRequest, forCreationAt date: Date? = nil) -> URL? {
+        guard let identifier = self.identifier(for: request) else { return nil }
+        guard let directory = cacheDirectory else { return nil }
+        
+        var foundURL: URL?
+        
+        enumerateDiskEntries { (entry, stop) in
+            if entry.identifier == identifier {
+                foundURL = entry.url
+                stop = true
+            }
+        }
+        
+        if let date = date {
+            // If we're trying to _create_ an entry and it already exists, then we can't -- we should evict the old one first.
+            if foundURL != nil {
+                return nil
```

Comment:
@drodriguez I cannot _guarantee_ that this check will not be executed even around L491, because we may be racing. You're correct that a behavior guarantee would be easier to achieve with filesystem locking, but as noted we may be operating without and not have a choice in the matter.

I need to sit down and think more deeply about the date being part of the filename, since it may ruin what little atomicity we have in `store…` and lead to duplicate entries for the same URL. It may be that we can just blindly store and then do an eviction pass at the end for all URLs that have the same key and are not _the highest timestamp we find on disk_ (not the one we just stored!), but I need to think through the racing implications of that.

Reply:
Yes, my comments about locking the file storage wasn't about locking it against two processes accessing the same cache, but about the same process accessing its own cache from two threads. My original design serialized many of the access to disk to avoid those kind of problems. Knowing that the access was serial simplified many of the checks a lot.

Reply:
The problem with that reasoning is, given that `URLCache.shared` is the default, it is not useful to do that anyway — the moment a second process runs on the same node it will use the same path. So we need to be flexible enough to work in that case anyway.

Reply:
Two processes of the same binary will create problems with the file system storage. That's unavoidable without file level locking. However I don’t think the situation happens that often for consumer software (different story for server software, though). My intention providing in-process serialization was to honor the documentation that states that `URLCache` is thread safe.

Reply:
The moment we have to reason about safety at an interprocess level without locking, if we get it right, we by definition have safety at an intra-process, inter-thread level without locking. At that point, adding locking is at best a performance issue and at worst hides interprocess issues that we would've caught otherwise.